File: 프로그래머스/3/17676. ［1차］ 추석 트래픽/［1차］ 추석 트래픽.py
def solution(lines):
    intervals = []  # (start_ms, end_ms)

    for line in lines:
        _, time_str, dur_str = line.split()

        end_ms = parse_time_to_ms(time_str)
        dur_ms = parse_duration_to_ms(dur_str)
        start_ms = end_ms - dur_ms + 1  # 시작/끝 포함

        intervals.append((start_ms, end_ms))

    answer = 0

    # 각 로그의 시작/끝 시점을 기준으로 1초 구간 [t, t+999] 검사
    for start, end in intervals:
        for window_start in (start, end):
            window_end = window_start + 999
            count = 0

            for s, e in intervals:
                # 구간이 겹치면 카운트
                if e >= window_start and s <= window_end:
                    count += 1

            answer = max(answer, count)

    return answer


def parse_time_to_ms(time_str):
    # "hh:mm:ss.sss"
    h = int(time_str[0:2])
    m = int(time_str[3:5])
    s = int(time_str[6:8])
    ms = int(time_str[9:12])

    return ((h * 3600 + m * 60 + s) * 1000) + ms


def parse_duration_to_ms(dur_str):
    # "2s", "2.0s", "0.351s"
    t = dur_str[:-1]  # remove 's'

    if '.' not in t:
        return int(t) * 1000

    sec_str, frac_str = t.split('.')
    frac_str = (frac_str + "000")[:3]  # 소수부 3자리 맞추기

    return int(sec_str) * 1000 + int(frac_str)

File: 프로그래머스/3/17676. ［1차］ 추석 트래픽/［1차］ 추석 트래픽.js
function solution(lines) {
    const intervals = []; // [startMs, endMs]

    for (const line of lines) {
        const parts = line.split(" ");
        const timeStr = parts[1];
        const durStr = parts[2];

        const endMs = parseTimeToMs(timeStr);
        const durMs = parseDurationToMs(durStr);
        const startMs = endMs - durMs + 1; // 시작/끝 포함

        intervals.push([startMs, endMs]);
    }

    let answer = 0;

    // 각 로그의 시작/끝 시점을 기준으로 1초 구간 [t, t+999] 검사
    for (const [start, end] of intervals) {
        const candidates = [start, end];

        for (const windowStart of candidates) {
            const windowEnd = windowStart + 999;
            let count = 0;

            for (const [s, e] of intervals) {
                // 구간 [s, e] 와 [windowStart, windowEnd]가 겹치면 카운트
                if (e >= windowStart && s <= windowEnd) {
                    count++;
                }
            }

            if (count > answer) answer = count;
        }
    }

    return answer;
}

function parseTimeToMs(timeStr) {
    // "hh:mm:ss.sss"
    const h = Number(timeStr.slice(0, 2));
    const m = Number(timeStr.slice(3, 5));
    const s = Number(timeStr.slice(6, 8));
    const ms = Number(timeStr.slice(9, 12));

    return ((h * 3600 + m * 60 + s) * 1000) + ms;
}

function parseDurationToMs(durStr) {
    // "2s", "2.0s", "0.351s"
    const t = durStr.slice(0, -1); // remove 's'

    if (!t.includes(".")) {
        return Number(t) * 1000;
    }

    const [secStr, fracStrRaw] = t.split(".");
    let fracStr = fracStrRaw;

    while (fracStr.length < 3) fracStr += "0";
    if (fracStr.length > 3) fracStr = fracStr.slice(0, 3);

    return Number(secStr) * 1000 + Number(fracStr);
}